http/common/quic.go

```diff
@@ -0,0 +1,196 @@
+	if err != nil {
+		return nil, err
+	}
+	stream, err := conn.AcceptStream(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &QuicConn{Conn: conn, Stream: stream}, nil
+}
+
+func (l *QuicListener) Addr() net.Addr {
+	return l.Listener.Addr()
+}
+
+func (l *QuicListener) Close() error {
+	return l.Listener.Close()
+}
+
+// Setup a bare-bones TLS config for the server
+func GenerateTLSConfig() *tls.Config {
+	pubKey, privKey, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		panic(err)
+	}
+	template := x509.Certificate{SerialNumber: big.NewInt(1)}
+	certDER, err := x509.CreateCertificate(rand.Reader, &template, &template, pubKey, privKey)
+	if err != nil {
+		panic(err)
+	}
+	pkb, err := x509.MarshalPKCS8PrivateKey(privKey)
+	if err != nil {
+		panic(err)
+	}
+	keyPEM := pem.EncodeToMemory(&pem.Block{Type: "ED25519 PRIVATE KEY", Bytes: pkb})
+	certPEM := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: certDER})
+
+	tlsCert, err := tls.X509KeyPair(certPEM, keyPEM)
+	if err != nil {
+		panic(err)
+	}
+	// ALPN (NextProtos) is externally visible as part of the QUIC TLS
+	// handshake, in the client/server hello, so pick a common protocol
+	// rather than something uniquely fingerprintable to katzenpost.
+	return &tls.Config{Certificates: []tls.Certificate{tlsCert}, NextProtos: []string{http3.NextProtoH3}}
+}
+
+func DialURL(u *url.URL, ctx context.Context, dialFn func(ctx context.Context, network, address string) (net.Conn, error)) (net.Conn, error) {
+	switch u.Scheme {
+	case "tcp", "tcp4", "tcp6":
+		// XXX: make sure to use the supplied dialer for proxy users
+		conn, err := dialFn(ctx, "tcp", u.Host)
+		if err != nil {
+			return nil, err
+		} else {
+			return conn, nil
+		}
+	case "quic":
+		// http/3 quic connector
+		// XXX: will need to add the TLS certificate
+		// fingerprint to the authority configuration
+		// or obtain valid CA-signed certificates for
+		// the authorities.
+		pool, err := x509.SystemCertPool()
+		if err != nil {
+			panic(err)
+		}
+		tlsConf := &tls.Config{
+			RootCAs:            pool,
+			InsecureSkipVerify: true, // XXX
+			// ALPN is externally visible as part of the client/server hello,
+			// so pick a common protocol rather than something fingerprintable.
+			NextProtos: []string{http3.NextProtoH3},
+		}
+
+		// XXX if an UpstreamProxy is specified, this dial function will fail,
+		// which is what we want
+
+		// TODO: Write SOCKS5 Dialer that wraps quic.Dial
+		// TODO: choose a socks5 client library that supports socks5 UDP Associate
+		// TODO: support QUIC Datagram HTTP Connect proxy dialers
+		_, err = dialFn(ctx, "udp", u.Host)
+		if err != nil {
+			return nil, err
+		}
+		// However the default net.Dialer returns a connected socket that
+		// makes quic.Dial fails with error: use of WriteTo with pre-connected UDP ..
+		// because we need to bind the local listening socket with net.ListenUDP.
+
+		// XXX: verify that quic.Dial does not leak DNS
```

Comment:
can we add a ticket link here? and your name
// TODO(masala): verify ... and ticket # blah